bot/helper/reqHelper.ts

```diff
@@ -0,0 +1,202 @@
+import { DefaultLineSetting } from "./constant";
+
+/**
+ * 
+ * @param url GitHub-related url
+ * @param useGithubRendered option to choose whether to get rendered HTML directly from GitHub
+ * @returns 
+ */
+export async function reqCodeDataFromGitHubAPI(url: string, useGithubRendered: boolean){
+    // Typically, An GitHub code permalink URL will be 
+    // `https://github.com/OfficeDev/TeamsFx-Samples/blob/master/test.py#L1-L6?x=a&y=b` 
+    // Or `https://github.com/OfficeDev/TeamsFx-Samples/blob/master/test.py?x=a&y=b`
+    if (url.includes("?")){
+        url = url.split("?")[0];
+    }
+    let fileurl = url;
+    let startLine = DefaultLineSetting.defaultStartLine;
+    let endLine = DefaultLineSetting.defaultEndLine;
+    // If the Url specified the startLine and endLine
+    if (url.includes("#")){
+        const segmentListSharp = url.split("#");
+        const lines = segmentListSharp[1];
+        const lineNumList = lines.replace(/L/g,"").split("-");
+        // Get startLine and endLine
+        startLine = Number(lineNumList[0]) -1;
+        endLine = Number(lineNumList[1]) -1;  
+        fileurl = segmentListSharp[0]
+    }
+    const segmentList = fileurl.replace("https://github.com/","").replace("blob/","").split("/");
+    // name space is the first element of the segmentList.
+    const namespace = segmentList.shift();
+    // repo name is the second element of the segmentList.
+    const repoName = segmentList.shift();
+    // ref is the third part of the segmentList, ref means the name of the commit/branch/tag.
+    const ref = segmentList.shift();
+    // connect the remaining part to get file path.
+    const path = segmentList.reduce((a,b) => a + "/" + b);
+   
+    var content: string;
+    // Use third-part API to render code to HTML.
+    if (!useGithubRendered){
+        // Request whole content from GitHub API with metadata.
+        const rawContent = await reqInfoFromGitHubAPI(namespace, repoName, path, ref);
+        const contentToRender = segmentContent(rawContent, startLine, endLine);
+        const language = getLanguage(path);
+        content = await renderContent(language, contentToRender);
+    }
+    // Directly use GitHub-rendered HTML of the file.
+    // Notice: No syntax-highlighting; Hard to get certain lines of code from rendered HTML.
+    else {
+        content = await reqRenderedHTMLFromGitHub(namespace, repoName, path);
+    }
+
+    var card = new CodeCard(`[${ref}]${namespace}/${repoName}/${path}: Lines ${startLine+1} - ${endLine+1}`, content, url);
+    return card;
+}
+
+/**
+ * Function to request file data from GitHub API.
+ * @param namespace namespace
+ * @param repoName name of the repository
+ * @param path path of the file
+ * @param ref name of the commit/branch/tag
+ * @returns raw content string of the file
+ */
+ async function reqInfoFromGitHubAPI(namespace: string, repoName: string, path: string, ref: string){
+    if (ref == undefined){
```

Comment:
Same here, the convention is to use ```if (!ref)``` for comparing with undefined.

Reply:
Fixed